### DISCUSSION TITLE
Generating Huawei device tree

### DISCUSSION THREAD

**Discussion by ghost** [SebaUbuntu/TWRP-device-tree-generator#66] [Q&A]
Hi everyone, I am trying to generate device tree for Huawei P10 lite using the original firmware (downloaded from  [https://huaweidl.com/download/p10-lite/was-lx1/8-0-0-394/](https://huaweidl.com/download/p10-lite/was-lx1/8-0-0-394/) ). There is a screenshot of all images available in the stock firmware downloaded [https://ibb.co/SfX0z0V](https://ibb.co/SfX0z0V).

I have tried to use the `erecovery` images without success (because there is no `recovery_kernel.img` in the files). The exact command is below.
```
python3 -m  twrpdtgen --no-git --huawei --recovery_ramdisk erecovery_ramdisk.img --recovery_vendor erecovery_vendor.img --recovery_kernel erecovery_kernel.img 
```

After invoking it, I got the following error:

```
Traceback (most recent call last):
  File "/usr/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/home/jakub/.local/lib/python3.8/site-packages/twrpdtgen/__main__.py", line 52, in <module>
    dt = DeviceTree(args.output, recovery_image=args.recovery_image,
  File "/home/jakub/.local/lib/python3.8/site-packages/twrpdtgen/device_tree.py", line 53, in __init__
    aik = huawei_utils.extract()
  File "/home/jakub/.local/lib/python3.8/site-packages/twrpdtgen/utils/huawei.py", line 40, in extract
    self.recovery_kernel_aik.extract(self.recovery_kernel)
  File "/home/jakub/.local/lib/python3.8/site-packages/twrpdtgen/utils/aik_manager.py", line 81, in extract
    raise RuntimeError(f"AIK extraction failed, return code {returncode}")
RuntimeError: AIK extraction failed, return code 1
```

I looked into the code and find out, that Android Image Kitchen `unpackimg.sh` script is called on the images provided.
But the `unpackimg.sh` script fails on every image from the list (image linked above). Which images should be provided to the script to work properly? Thanks in advance.

**Comment by SebaUbuntu** [comment]
Try to send all 3 images here, extraction should work fine

**Reply by ghost** [reply]
Thanks for a fast reply. There are images used.
[https://www.filemail.com/d/alsdyvskxuzzebd](https://www.filemail.com/d/alsdyvskxuzzebd)

**Reply by SebaUbuntu** [reply]
So strange, none of these images matches an AOSP image magic header, and I've tested several Huawei recovery images
Please be sure to properly extract images, DON'T USE 7ZIP OR ANY GENERIC UTILITY, you must use this https://forum.xda-developers.com/t/tool-huawei-update-extractor-updated-v0-9-9-5-open-source-library.2433454/

**Reply by ghost** [reply]
Great, that was the problem. Now only some information is missing in the images (codename, brand, model), but these can be written in manually with the AIK.

**Comment by ghost** [comment]
After successfully extraction the images, I get error like these
```
TWRP device tree generator
Version 1.3.0

FIXME: Images are check - ok
[INFO] Cloning AIK...
[INFO] Cloning AIK...
[INFO] Cloning AIK...
Extraction is in progress /tmp/tmpdcnkvzrs with /tmp/tmpdcnkvzrs/recovery.img
Extraction is in progress /tmp/tmpjmbo0l9u with /tmp/tmpjmbo0l9u/recovery.img
Extraction is in progress /tmp/tmpapo6700i with /tmp/tmpapo6700i/recovery.img
FIXME: AIK done -ok
Traceback (most recent call last):
  File "/usr/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/home/jakub/.local/lib/python3.8/site-packages/twrpdtgen/__main__.py", line 52, in <module>
    dt = DeviceTree(args.output, recovery_image=args.recovery_image,
  File "/home/jakub/.local/lib/python3.8/site-packages/twrpdtgen/device_tree.py", line 64, in __init__
    self.build_prop_reader = BuildPropReader(self.build_prop)
  File "/home/jakub/.local/lib/python3.8/site-packages/twrpdtgen/info_extractors/buildprop.py", line 38, in __init__
    self.model = self.get_prop(DEVICE_MODEL, "model")
  File "/home/jakub/.local/lib/python3.8/site-packages/twrpdtgen/info_extractors/buildprop.py", line 67, in get_prop
    raise AssertionError(f"Device {error} could not be found in build.prop")
AssertionError: Device model could not be found in build.prop
```
After fixing them using the AIK (unpacking then writting the missing info in `prop.default` and repacking image). The device tree is generated. Also the twrp compilation is successfull, but after flashing I got the following error during the boot process
```
ERROR MODE

Attention!
Please update system again.

Error!
Func NO: 11 (recovery image)
Error NO: 2 (load failed)
```

Do you have any idea, what the issue might be?
